File: examples/companion-packages/dotnet-csharp-lib/src/AgentXM.Examples.TinyFlags.CSharp/TinyFlags.cs
using System.Collections.Frozen;
using System.Collections.Immutable;
using System.Diagnostics;

namespace AgentXM.Examples.TinyFlags.CSharp;

public sealed record EvaluationContext(
    string? UserId = null,
    string? AccountId = null,
    string? SessionId = null);

public abstract record FlagDefinition
{
    private protected FlagDefinition() { }
}

public sealed record BooleanFlagDefinition : FlagDefinition
{
    internal BooleanFlagDefinition(bool @default, int? rollout)
    {
        Default = @default;
        Rollout = rollout;
    }

    public bool Default { get; }
    public int? Rollout { get; }
}

public sealed record VariantFlagDefinition : FlagDefinition
{
    internal VariantFlagDefinition(
        ImmutableArray<string> variants,
        string @default,
        FrozenDictionary<string, int>? rollout)
    {
        Variants = variants;
        Default = @default;
        Rollout = rollout;
    }

    public ImmutableArray<string> Variants { get; }
    public string Default { get; }
    public FrozenDictionary<string, int>? Rollout { get; }
}

public abstract record FlagValue
{
    private FlagValue() { }

    public sealed record Bool(bool Value) : FlagValue;

    public sealed record Variant(string Value) : FlagValue;
}

public static class TinyFlag
{
    public static BooleanFlagDefinition Boolean(bool defaultValue = false, int? rollout = null) =>
        new(defaultValue, NormalizePercentage(rollout, nameof(rollout)));

    public static VariantFlagDefinition Variant(
        ReadOnlySpan<string> variants,
        string? defaultValue = null,
        IReadOnlyDictionary<string, int>? rollout = null)
    {
        if (variants.IsEmpty)
        {
            throw new ArgumentException("Variant flags require at least one variant.", nameof(variants));
        }

        var unique = new HashSet<string>(StringComparer.Ordinal);
        foreach (var variant in variants)
        {
            if (variant.Length == 0 || !unique.Add(variant))
            {
                throw new ArgumentException("Variant names must be unique non-empty strings.", nameof(variants));
            }
        }

        var resolvedDefault = defaultValue ?? variants[0];
        if (!unique.Contains(resolvedDefault))
        {
            throw new ArgumentException("Variant default must be one of the variants.", nameof(defaultValue));
        }

        return new VariantFlagDefinition(
            [.. variants],
            resolvedDefault,
            NormalizeVariantRollout(rollout, unique.ToFrozenSet(StringComparer.Ordinal)));
    }

    private static int? NormalizePercentage(int? value, string label) => value switch
    {
        null => null,
        < 0 or > 100 => throw new ArgumentOutOfRangeException(label, "Percentage must be from 0 to 100."),
        _ => value,
    };

    private static FrozenDictionary<string, int>? NormalizeVariantRollout(
        IReadOnlyDictionary<string, int>? rollout,
        FrozenSet<string> variants)
    {
        if (rollout is null)
        {
            return null;
        }

        var normalized = new Dictionary<string, int>(StringComparer.Ordinal);
        var total = 0;

        foreach (var (variant, percentage) in rollout)
        {
            if (!variants.Contains(variant))
            {
                throw new ArgumentException($"Rollout references unknown variant: {variant}.", nameof(rollout));
            }

            if (NormalizePercentage(percentage, $"rollout for '{variant}'") is not { } pct)
            {
                continue;
            }

            normalized[variant] = pct;
            total += pct;
        }

        if (total > 100)
        {
            throw new ArgumentOutOfRangeException(nameof(rollout), "Variant rollout percentages cannot exceed 100.");
        }

        return normalized.ToFrozenDictionary(StringComparer.Ordinal);
    }
}

public sealed class TinyFlags
{
    private readonly FrozenDictionary<string, FlagDefinition> definitions;

    private TinyFlags(FrozenDictionary<string, FlagDefinition> definitions)
    {
        this.definitions = definitions;
    }

    public IReadOnlyDictionary<string, FlagDefinition> Definitions => definitions;

    public static TinyFlags Create(IReadOnlyDictionary<string, FlagDefinition> definitions) =>
        new(definitions.ToFrozenDictionary(StringComparer.Ordinal));

    public bool Enabled(string name, EvaluationContext? context = null) =>
        Require<BooleanFlagDefinition>(name) switch
        {
            { Rollout: null } flag => flag.Default,
            { Rollout: { } rollout } => BucketFor(name, context) < rollout,
        };

    public string Variant(string name, EvaluationContext? context = null)
    {
        var flag = Require<VariantFlagDefinition>(name);
        if (flag.Rollout is null)
        {
            return flag.Default;
        }

        var bucket = BucketFor(name, context);
        var upper = 0;
        foreach (var (variant, percentage) in flag.Rollout)
        {
            upper += percentage;
            if (bucket < upper)
            {
                return variant;
            }
        }

        return flag.Default;
    }

    public FlagValue Evaluate(string name, EvaluationContext? context = null) =>
        RequireAny(name) switch
        {
            BooleanFlagDefinition => new FlagValue.Bool(Enabled(name, context)),
            VariantFlagDefinition => new FlagValue.Variant(Variant(name, context)),
            _ => throw new UnreachableException(),
        };

    private TFlag Require<TFlag>(string name) where TFlag : FlagDefinition =>
        RequireAny(name) switch
        {
            TFlag matched => matched,
            var other => throw new InvalidOperationException(
                $"TinyFlags flag '{name}' is not a {typeof(TFlag).Name} (got {other.GetType().Name})."),
        };

    private FlagDefinition RequireAny(string name) =>
        definitions.TryGetValue(name, out var flag)
            ? flag
            : throw new KeyNotFoundException($"Unknown TinyFlags flag: {name}.");

    private static int BucketFor(string name, EvaluationContext? context)
    {
        var key = context?.UserId ?? context?.AccountId ?? context?.SessionId ?? "anonymous";
        return (int)(HashString($"{name}:{key}") % 100);
    }

    private static uint HashString(string value)
    {
        unchecked
        {
            var hash = 2166136261u;
            foreach (var character in value)
            {
                hash ^= character;
                hash *= 16777619u;
            }

            return hash;
        }
    }
}

File: examples/companion-packages/dotnet-fsharp-nuget-lib/src/AgentXM.Examples.TinyFlags.FSharp/TinyFlags.fs
namespace AgentXM.Examples.TinyFlags.FSharp

open System
open System.Collections.Generic

/// Identity attributes used to bucket a flag evaluation deterministically.
type EvaluationContext =
    { UserId: string option
      AccountId: string option
      SessionId: string option }

module EvaluationContext =
    /// An evaluation context with no identity attributes.
    let empty =
        { UserId = None
          AccountId = None
          SessionId = None }

/// A validated flag definition produced by `FlagDefinition.booleanFlag` or `variantFlag`.
type FlagDefinition =
    | BooleanFlag of defaultValue: bool * rollout: int option
    | VariantFlag of variants: string list * defaultValue: string * rollout: Map<string, int> option

/// The typed result of evaluating a flag.
type FlagValue =
    | BoolValue of bool
    | VariantValue of string

/// Smart constructors for `FlagDefinition` that validate inputs and accept
/// optional named parameters. Static-only — never instantiated.
[<AbstractClass; Sealed>]
type Flag private () =

    static member private ValidatePercentage(label, percentage) =
        if percentage < 0 || percentage > 100 then
            invalidArg label $"Percentage must be from 0 to 100; received {percentage}."

        percentage

    static member private NormalizeRollout(variants: string list, allocation: Map<string, int>) =
        let known = Set.ofList variants

        let normalized, total =
            allocation
            |> Map.fold
                (fun (acc, runningTotal) variant percentage ->
                    if not (Set.contains variant known) then
                        invalidArg "rollout" $"Rollout references unknown variant: {variant}."

                    let validated = Flag.ValidatePercentage($"rollout for '{variant}'", percentage)
                    Map.add variant validated acc, runningTotal + validated)
                (Map.empty, 0)

        if total > 100 then
            invalidArg "rollout" $"Variant rollout percentages cannot exceed 100; received {total}."

        normalized

    /// Define a boolean feature flag. `rollout` is an optional 0–100 enabled percentage.
    static member Boolean(defaultValue: bool, ?rollout: int) : FlagDefinition =
        BooleanFlag(defaultValue, rollout |> Option.map (fun pct -> Flag.ValidatePercentage("rollout", pct)))

    /// Define a multi-variant feature flag. `defaultValue` defaults to the first variant.
    static member Variant
        (variants: string list, ?defaultValue: string, ?rollout: Map<string, int>)
        : FlagDefinition =
        match variants with
        | [] -> invalidArg "variants" "Variant flags require at least one variant."
        | _ ->
            let unique = List.distinct variants

            if unique.Length <> variants.Length || List.exists String.IsNullOrEmpty unique then
                invalidArg "variants" "Variant names must be unique non-empty strings."

            let chosenDefault = defaultArg defaultValue variants.Head

            if not (List.contains chosenDefault unique) then
                invalidArg "defaultValue" $"Variant default '{chosenDefault}' is not one of the variants."

            VariantFlag(unique, chosenDefault, rollout |> Option.map (fun r -> Flag.NormalizeRollout(unique, r)))

/// A bundle of named flag definitions. Construct with `TinyFlags.create`.
type TinyFlags = private { Definitions: Map<string, FlagDefinition> }

module TinyFlags =

    let private fnv1a (value: string) =
        let mutable hash = 2166136261u

        for character in value do
            hash <- hash ^^^ uint32 character
            hash <- hash * 16777619u

        hash

    let private bucketFor (name: string) (context: EvaluationContext) =
        let key =
            context.UserId
            |> Option.orElse context.AccountId
            |> Option.orElse context.SessionId
            |> Option.defaultValue "anonymous"

        int (fnv1a $"{name}:{key}" % 100u)

    let private require name flags =
        match Map.tryFind name flags.Definitions with
        | Some definition -> definition
        | None -> raise (KeyNotFoundException $"Unknown TinyFlags flag: {name}.")

    /// Build a TinyFlags bundle from a sequence of (name, definition) pairs.
    let create (definitions: seq<string * FlagDefinition>) =
        { Definitions = Map.ofSeq definitions }

    /// Evaluate a boolean flag. Throws if the flag is not boolean.
    let enabled (name: string) (context: EvaluationContext) flags =
        match require name flags with
        | BooleanFlag(defaultValue, None) -> defaultValue
        | BooleanFlag(_, Some rollout) -> bucketFor name context < rollout
        | VariantFlag _ -> invalidOp $"TinyFlags flag '{name}' is not a boolean flag."

    /// Evaluate a variant flag. Throws if the flag is not a variant flag.
    let variant (name: string) (context: EvaluationContext) flags =
        match require name flags with
        | VariantFlag(_, defaultValue, None) -> defaultValue
        | VariantFlag(_, defaultValue, Some rollout) ->
            let bucket = bucketFor name context

            let chosen, _ =
                rollout
                |> Map.toSeq
                |> Seq.fold
                    (fun (selected, upperBound) (candidate, percentage) ->
                        match selected with
                        | Some _ -> selected, upperBound
                        | None ->
                            let nextBound = upperBound + percentage

                            if bucket < nextBound then
                                Some candidate, nextBound
                            else
                                None, nextBound)
                    (None, 0)

            chosen |> Option.defaultValue defaultValue
        | BooleanFlag _ -> invalidOp $"TinyFlags flag '{name}' is not a variant flag."

    /// Evaluate any flag, returning a typed `FlagValue`.
    let evaluate (name: string) (context: EvaluationContext) flags =
        match require name flags with
        | BooleanFlag _ -> BoolValue(enabled name context flags)
        | VariantFlag _ -> VariantValue(variant name context flags)
